File: problem-solving/largest-sum/largest-sum.ts
export function findThreeLargestNumbers(array: number[]) {
  if (array.length === 3) return array;
  const xs = [...array].sort((a, b) => a - b);

  return xs.slice(xs.length - 3);
}

export function findThreeLargestNumbers2(array: number[]) {
  const highestNumberOfThree: Array<number | null> = [null, null, null];

  for (const n of array) {
    updateList(highestNumberOfThree, n);
  }

  return highestNumberOfThree;
}

function updateList(arr: Array<number | null>, n: number) {
  if (arr[2] === null || n > arr[2]) {
    change(arr, n, 2);
  } else if (arr[1] === null || n > arr[1]) {
    change(arr, n, 1);
  } else if (arr[0] === null || n > arr[0]) {
    change(arr, n, 0);
  }
}

function change(list: Array<number | null>, n: number, index: number) {
  for (let i = 0; i <= index; i++) {
    if (index === i) {
      list[i] = n;
    } else {
      list[i] = list[i + 1];
    }
  }
}

const xs = [141, 1, 17, -7, -17, -27, 18, 541, 8, 7, 7];
console.log(findThreeLargestNumbers2(xs)); //[18,141,541]

File: problem-solving/largest-sum/largest-sum.go
package main

import (
	"fmt"
	"math"
	"sort"
)

// FindThreeLargestNumbers fn
func FindThreeLargestNumbers(array []int) []int {
	xs := array
	sort.Ints(xs)
	return xs[len(xs)-3:]
}

// FindThreeLargestNumbers2 fn withour sorting
func FindThreeLargestNumbers2(array []int) []int {
	three := []int{math.MinInt32, math.MinInt32, math.MinInt32}
	for _, n := range array {
		update(three, n)
	}

	return three
}

func update(array []int, n int) {
	if n > array[2] {
		shift(array, n, 2)
	} else if n > array[1] {
		shift(array, n, 1)
	} else if n > array[0] {
		shift(array, n, 0)
	}
}

func shift(array []int, n, index int) {
	for i := 0; i <= index; i++ {
		if i == index {
			array[i] = n
		} else {
			array[i] = array[i+1]
		}
	}
}

func main() {
	xs := []int{-1, -2, -3, -7, -17, -27, -18, -541, -8, -7, 7}

	fmt.Println(FindThreeLargestNumbers2(xs))
}
